### PR TITLE
Replace caml_int_compare and caml_float_compare with primitives

### PR DESCRIPTION
**Pull request by gretay-js**
Add new compiler primitives to implement comparison operations that return int (not bool) result. These operations are important for IntMap.
Instead of calls to `caml_int*_compare` and `caml_float_compare`, emit native code directly.  This enables some optimizations when arguments of compare operations are statically known constants.  There is no change to machine-specific code generation, because the new primitives are translated in `cmmgen` to existing cmm constructs.

Recently, PR #2250 @smuenzel-js   proposed to improve these C functions.  It will still be useful for bytecode and int64 on 32-bit targets.  For native code, the current PR gives better performance than PR #2250.

Benchmarking on amd64 target Intel(R) Xeon(R) Gold 6144 on two microbenchmarks:

test5.ml: Tight loop example from PR #2250
```
┌───────────────────┬─────────────────┬─────────────────┬───────┬────────────┬─────────────┬───────┐
│                   │    instructions │      Difference │     % │ clock (ms) │  Difference │     % │
├───────────────────┼─────────────────┼─────────────────┼───────┼────────────┼─────────────┼───────┤
│    test5-base.exe │ 280,092,029,458 │               0 │ 1.000 │ 29,386.268 │           0 │ 1.000 │
│  test5-pr2250.exe │ 268,091,554,395 │ -12,000,475,063 │ 0.957 │ 29,439.480 │      53.212 │ 1.002 │
│ test5-compare.exe │ 220,056,326,489 │ -60,035,702,969 │ 0.786 │ 17,720.873 │ -11,665.395 │ 0.603 │
└───────────────────┴─────────────────┴─────────────────┴───────┴────────────┴─────────────┴───────┘
```

test6.ml: IntMap loop that calls add, remove, find, and mem.
```
┌───────────────────┬────────────────┬────────────────┬───────┬────────────┬────────────┬───────┐
│                   │   instructions │     Difference │     % │ clock (ms) │ Difference │     % │
├───────────────────┼────────────────┼────────────────┼───────┼────────────┼────────────┼───────┤
│    test6-base.exe │ 72,942,680,772 │              0 │ 1.000 │ 13,188.722 │          0 │ 1.000 │
│  test6-pr2250.exe │ 72,034,245,520 │   -908,435,252 │ 0.988 │ 13,126.140 │    -62.582 │ 0.995 │
│ test6-compare.exe │ 68,470,617,181 │ -4,472,063,591 │ 0.939 │ 12,640.141 │   -548.581 │ 0.958 │
└───────────────────┴────────────────┴────────────────┴───────┴────────────┴────────────┴───────┘
```

Inria's precheck all tests pass (build # 206).

### CODE REVIEW

File: asmcomp/cmmgen.ml [hunk 1]
```diff
@@ -2421,6 +2441,23 @@ and transl_prim_2 env p arg1 arg2 dbg =
...
+      let a1 = transl_unbox_int dbg env bi arg1 in
+      let a2 = transl_unbox_int dbg env bi arg2 in
+      mk_compare_ints dbg a1 a2
+  | Pcompare_floats ->
+      let a1 = transl_unbox_float dbg env arg1 in
+      let a2 = transl_unbox_float dbg env arg2 in
+      let op1 = Cop(Ccmpf(CFgt), [a1; a2], dbg) in
+      let op2 = Cop(Ccmpf(CFlt), [a1; a2], dbg) in
+      let op3 = Cop(Ccmpf(CFeq), [a1; a1], dbg) in
+      let op4 = Cop(Ccmpf(CFeq), [a2; a2], dbg) in
```
Comment (stedolan) [inline asmcomp/cmmgen.ml:924]: This looks correct, but it took me a while to puzzle it out and a comment would help!

(My understanding is: if either operand is `nan`, then `op1 = op2 = 0` and `op3 - op4` orders `nan` before non-`nan` values. If neither operand is `nan`, then `op3 - op4` is `0` (both are `1`), so `op1 - op2` gives the answer)

**Comment by smuenzel-js** [comment]
We probably just need to rebase to fix the checks (I've done that here, if you want to pull: https://github.com/smuenzel-js/ocaml/tree/compare-primitive) and add some comments in the section @stedolan mentioned, and then this should be ready for review, right?

**Comment by gretay-js** [comment]
Rebased and updated the comment.
Thanks @stedolan for the review and @smuenzel-js for help!

**Comment by gretay-js** [comment]
Ping... there wasn't any objection to this being merged, right?

**Comment by smuenzel-js** [comment]
@stedolan : ping

**Comment by smuenzel** [comment]
one more ping :)

File: asmcomp/cmm_helpers.ml [hunk 2]
```diff
@@ -274,6 +274,24 @@ let mk_not dbg cmm =
...
+  | Cconst_int (c1, _), Cconst_int (c2, _) -> begin
+      if c1 = c2 then int_const dbg 0
+      else if c1 < c2 then int_const dbg (-1)
+      else int_const dbg 1
+    end
+  | Cconst_natint (c1, _), Cconst_natint (c2, _) -> begin
+      if c1 = c2 then int_const dbg 0
+      else if c1 < c2 then int_const dbg (-1)
+      else int_const dbg 1
+    end
```
Comment (lthls) [inline asmcomp/cmm_helpers.ml]: I think you could handle the cases where `a1` and `a2` are `Cconst_int` and `Cconst_natint` too.

File: asmcomp/cmmgen.ml [hunk 3]
```diff
@@ -907,6 +908,30 @@ and transl_prim_2 env p arg1 arg2 dbg =
       asr_int_caml (transl env arg1) (transl env arg2) dbg
   | Pintcomp cmp ->
       int_comp_caml cmp (transl env arg1) (transl env arg2) dbg
+  | Pcompare_ints ->
+      (* Compare directly on tagged ints *)
+      mk_compare_ints dbg (transl env arg1) (transl env arg2)
+  | Pcompare_bints bi ->
+      let a1 = transl_unbox_int dbg env bi arg1 in
+      let a2 = transl_unbox_int dbg env bi arg2 in
+      mk_compare_ints dbg a1 a2
```
Comment (lthls) [inline asmcomp/cmmgen.ml:917]: I think this code is wrong: on 32-bit architectures, 64-bit integers are not translated to machine integers, but to pairs of integers. I'm not sure exactly what will happen, hopefully a type mismatch error later in the compilation pipeline.

File: asmcomp/cmmgen.ml [hunk 4]
```diff
@@ -907,6 +908,30 @@ and transl_prim_2 env p arg1 arg2 dbg =
...
+      let op4 = Cop(Ccmpf(CFeq), [a2; a2], dbg) in
+      (* If both operands a1 and a2 are not NaN, then op3 = op4 = 1,
+         and the result is op1 - op2.
+         If at least one of the operands is NaN,
+         then op1 = op2 = 0, and the result is op3 - op4,
+         which orders NaN before other values.
+         To detect if the operand is NaN, we use the property:
+         for all x, NaN is not equal to x, even if x is NaN.
+         Therefore, op3 is 0 if and only if a1 is NaN,
+         and op4 is 0 if and only if a2 is NaN. *)
```
Comment (lthls) [inline asmcomp/cmmgen.ml]: There's a nice comment in `runtime/floats.c`, on function `caml_float_compare_unboxed`, that explains the choice of operations. I'd suggest to add cross references at both places to keep them consistent with each other.

**Review by lthls — CHANGES_REQUESTED** [review]
The handling of 32-bit architectures needs to be fixed. In particular, there are a number of suspect failures in the CI that need to be looked at. You can probably reproduce the failures by configuring your compiler to use a 32-bit toolchain (see [this opam file](https://github.com/ocaml/opam-repository/blob/master/packages/ocaml-variants/ocaml-variants.4.09.0%2B32bit/opam) for the configure flags).

File: asmcomp/cmmgen.ml (hunk 3 above)
Comment (gretay-js) [reply asmcomp/cmmgen.ml:917]: Fixed. 
I have already had it working on another branch but somehow the fix was not included here after all the rebases, sorry.

File: asmcomp/cmmgen.ml (hunk 4 above)
Comment (gretay-js) [reply asmcomp/cmmgen.ml]: Done

File: asmcomp/cmm_helpers.ml (hunk 2 above)
Comment (gretay-js) [reply asmcomp/cmm_helpers.ml]: The code already handles the case of 
https://github.com/ocaml/ocaml/blob/040ec34dd1bab9b9ab64e492a94c03c8c906ebb0/asmcomp/cmm_helpers.ml#L279
and 
https://github.com/ocaml/ocaml/blob/040ec34dd1bab9b9ab64e492a94c03c8c906ebb0/asmcomp/cmm_helpers.ml#L284

Are there other combinations with Cconst_* that it should check ?

Comment (lthls) [reply asmcomp/cmm_helpers.ml]: I was referring to cases where one is `Cconst_int` and the other is `Cconst_natint`. They're compatible.

Comment (gretay-js) [reply asmcomp/cmm_helpers.ml]: Ah, they are! I've just  updated the code to handle all 4 cases uniformly.

**Review by lthls — APPROVED** [review]

**Comment by lthls** [comment]
I'm leaving the week-end for other developers to voice objections, and will merge on Monday unless somebody complains strongly enough.

**Comment by gretay-js** [comment]
@lthls Thank you!

**Comment by lthls** [comment]
I'll pretend we're still Monday.
Thanks @gretay-js for the contribution !

**Pull request merged**